sdks/go/pkg/beam/runners/universal/runnerlib/stage.go

```diff
@@ -17,27 +17,128 @@ package runnerlib
 
 import (
 	"context"
+	"io"
+	"os"
 	"time"
 
 	"github.com/apache/beam/sdks/go/pkg/beam/artifact"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx"
 	"github.com/apache/beam/sdks/go/pkg/beam/internal/errors"
 	jobpb "github.com/apache/beam/sdks/go/pkg/beam/model/jobmanagement_v1"
 	"github.com/apache/beam/sdks/go/pkg/beam/util/grpcx"
+	"google.golang.org/grpc"
 )
 
 // Stage stages the worker binary and any additional files to the given
 // artifact staging endpoint. It returns the retrieval token if successful.
-func Stage(ctx context.Context, id, endpoint, binary, st string, files ...artifact.KeyedFile) (retrievalToken string, err error) {
+func Stage(ctx context.Context, id, endpoint, binary, st string) (retrievalToken string, err error) {
 	ctx = grpcx.WriteWorkerID(ctx, id)
 	cc, err := grpcx.Dial(ctx, endpoint, 2*time.Minute)
 	if err != nil {
 		return "", errors.WithContext(err, "connecting to artifact service")
 	}
 	defer cc.Close()
 
+	err = StageViaPorableApi(ctx, cc, binary, st)
+
+	if err == nil {
+		return "", err
+	} else {
+		return StageViaLegacyApi(ctx, cc, binary, st)
+	}
+}
+
+func StageViaPorableApi(ctx context.Context, cc *grpc.ClientConn, binary, st string) error {
+	client := jobpb.NewArtifactStagingServiceClient(cc)
+
+	stream, err := client.ReverseArtifactRetrievalService(context.Background())
+	if err != nil {
+		return err
+	}
+
+	if err := stream.Send(&jobpb.ArtifactResponseWrapper{StagingToken: st}); err != nil {
+		return err
+	}
+
+	for {
+		in, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		switch request := in.Request.(type) {
+		case *jobpb.ArtifactRequestWrapper_ResolveArtifact:
+			err = stream.Send(&jobpb.ArtifactResponseWrapper{
+				Response: &jobpb.ArtifactResponseWrapper_ResolveArtifactResponse{
+					&jobpb.ResolveArtifactsResponse{
+						Replacements: request.ResolveArtifact.Artifacts,
+					},
+				}})
+			if err != nil {
+				return err
+			}
+
+		case *jobpb.ArtifactRequestWrapper_GetArtifact:
+			TypeUrn := request.GetArtifact.Artifact.TypeUrn
```

Comment:
While not incorrect, it's a bit odd to see a capitalized variable name in Go because a capital letter indicates the identifier is Exported from the package. 

From a readability standpoint the capital initial tells the reader that the scope of this variable extends beyond this package, which isn't the case for a local variable in function scope. At the definition line (here) this is fine, but at later uses, it would be easy to make the mistake, miamatchign assumptions.

MixedCaps are correct though.

Consider typeUrn instead.

As a second note: you can put the definition inline in the switch header

switch typeUrn := ... ; typeUrn {
case ...

which limits its scope to just the switch.